File: backend/internal/service/token_bank.go
package service

import (
	"context"
	"time"

	infraerrors "github.com/Wei-Shaw/sub2api/internal/pkg/errors"
)

var ErrRentalNotFound = infraerrors.NotFound("RENTAL_NOT_FOUND", "储蓄账号不存在")

// RentalAccount intentionally contains no credentials or consumer identity.
type RentalAccount struct {
	ID          int64   `json:"id"`
	OwnerUserID int64   `json:"owner_user_id"`
	Name        string  `json:"name"`
	Platform    string  `json:"platform"`
	Type        string  `json:"type"`
	Status      string  `json:"status"`
	Schedulable bool    `json:"schedulable"`
	Requests    int64   `json:"requests"`
	Tokens      int64   `json:"tokens"`
	Revenue     float64 `json:"revenue"`
}

type RentalRevenue struct {
	ID            int64     `json:"id"`
	AccountID     int64     `json:"account_id"`
	OwnerUserID   int64     `json:"owner_user_id"`
	Platform      string    `json:"platform"`
	Model         string    `json:"model"`
	BillingType   int8      `json:"billing_type"`
	InputTokens   int64     `json:"input_tokens"`
	OutputTokens  int64     `json:"output_tokens"`
	CacheTokens   int64     `json:"cache_tokens"`
	BillAmount    float64   `json:"bill_amount"`
	OwnerShareBPS int       `json:"owner_share_bps"`
	OwnerAmount   float64   `json:"owner_amount"`
	AdminAmount   float64   `json:"admin_amount"`
	CreatedAt     time.Time `json:"created_at"`
}

type RentalOverview struct {
	Accounts      []RentalAccount `json:"accounts"`
	TotalAccounts int             `json:"total_accounts"`
	TotalRevenue  float64         `json:"total_revenue"`
	TodayRevenue  float64         `json:"today_revenue"`
	AdminRevenue  float64         `json:"admin_revenue"`
}

type RentalRevenuePage struct {
	Items []RentalRevenue `json:"items"`
	Total int             `json:"total"`
}

type TokenBankRepository interface {
	Overview(context.Context, int64, string, string, int, int) (*RentalOverview, error)
	Revenue(context.Context, int64, int64, string, int, int) (*RentalRevenuePage, error)
}

File: frontend/src/api/tokenBank.ts
import { apiClient } from './client'

export interface RentalPolicy {
  id?: number
  platform: string
  group_id: number
  owner_share_bps: number
  admin_user_id: number
  enabled: boolean
  version?: number
}
export interface RentalAccount {
  id: number
  owner_user_id: number
  name: string
  platform: string
  type: string
  status: string
  rental_status: string
  schedulable: boolean
  requests: number
  tokens: number
  revenue: number
}
export interface RentalOverview {
  accounts: RentalAccount[]
  total_accounts: number
  total_revenue: number
  today_revenue: number
  admin_revenue: number
}
export interface RentalRevenue {
  id: number
  account_id: number
  owner_user_id: number
  platform: string
  model: string
  billing_type: number
  input_tokens: number
  output_tokens: number
  cache_tokens: number
  bill_amount: number
  owner_share_bps: number
  owner_amount: number
  admin_amount: number
  created_at: string
}
export interface RevenuePage {
  items: RentalRevenue[]
  total: number
}
export interface RentalInput {
  name: string
  platform: string
  api_key?: string
  account_id?: number
}
export interface RentalAuth {
  auth_url: string
  session_id: string
}

export const tokenBankAPI = {
  policies: async (admin: boolean) =>
    (
      await apiClient.get<RentalPolicy[]>(
        `/${admin ? 'admin' : 'user'}/token-bank/policies`
      )
    ).data,
  overview: async (admin: boolean, params: Record<string, unknown>) =>
    (
      await apiClient.get<RentalOverview>(
        `/${admin ? 'admin' : 'user'}/token-bank/accounts`,
        { params }
      )
    ).data,
  revenue: async (admin: boolean, params: Record<string, unknown>) =>
    (
      await apiClient.get<RevenuePage>(
        `/${admin ? 'admin' : 'user'}/token-bank/revenue`,
        { params }
      )
    ).data,
  savePolicy: async (policy: RentalPolicy) =>
    apiClient.put('/admin/token-bank/policies', policy),
  importAccount: async (input: RentalInput) =>
    apiClient.post('/user/token-bank/accounts', input),
  startOAuth: async (input: RentalInput) =>
    (await apiClient.post<RentalAuth>('/user/token-bank/oauth/start', input))
      .data,
  finishOAuth: async (session_id: string, code: string, state: string) =>
    apiClient.post('/user/token-bank/oauth/finish', {
      session_id,
      code,
      state
    }),
  setStatus: async (id: number, status: string) =>
    apiClient.patch(`/user/token-bank/accounts/${id}/status`, { status })
}

export function parseRentalCallback(
  value: string,
  state: string
): { code: string; state: string } {
  const input = value.trim()
  if (input.startsWith('http://') || input.startsWith('https://')) {
    const url = new URL(input)
    return {
      code: url.searchParams.get('code') || '',
      state: url.searchParams.get('state') || ''
    }
  }
  const parts = input.split('#')
  return { code: parts[0], state: parts[1] || state.trim() }
}
